src/main/java/com/arpnetworking/utility/RegexAndMapReplacer.java

```diff
@@ -0,0 +1,296 @@
+            final LinkedHashMap<String, Map<String, String>> variables,
+            final ImmutableList.Builder<String> variablesUsedBuilder) {
+        if (numeric) {
+            final int replaceGroup = Integer.parseInt(replaceToken);
+            return matcher.group(replaceGroup);
+        } else {
+            final int prefixSeparatorIndex = replaceToken.indexOf(':');
+            final String prefix;
+            final String variableName;
+            if (prefixSeparatorIndex == replaceToken.length() - 1) {
+                throw new IllegalArgumentException(
+                        String.format("found prefix in variable replacement, but no variable name found: '%s'",
+                        replaceToken));
+            }
+            if (prefixSeparatorIndex != -1) {
+                prefix = replaceToken.substring(0, prefixSeparatorIndex);
+                variableName = replaceToken.substring(prefixSeparatorIndex + 1);
+            } else {
+                prefix = "";
+                variableName = replaceToken;
+            }
+            if (prefix.isEmpty()) {
+                return replaceNonPrefix(matcher, variables, variablesUsedBuilder, variableName);
+
+            } else {
+                return replacePrefix(matcher, variables, variablesUsedBuilder, prefix, variableName);
+            }
+        }
+    }
+
+    private static String replacePrefix(
+            final Matcher matcher,
+            final LinkedHashMap<String, Map<String, String>> variables,
+            final ImmutableList.Builder<String> variablesUsedBuilder,
+            final String prefix,
+            final String variableName) {
+        if (prefix.equals("capture")) {
+            if (isNumeric(variableName)) {
+                final Integer groupIndex = Integer.valueOf(variableName);
+                return matcher.group(groupIndex);
+            } else {
+                try {
+                    return matcher.group(variableName);
+                } catch (final IllegalArgumentException ignored2) { // No group with this name
+                    return "";
+                }
+            }
+        }
+
+        // Only record variables that are not captures
+        variablesUsedBuilder.add(String.format("%s:%s", prefix, variableName));
+
+        final Map<String, String> variableMap = variables.get(prefix);
+        if (variableMap == null) {
+            throw new IllegalArgumentException(String.format("could not find map for variables with prefix '%s'", prefix));
+        }
+        return variableMap.getOrDefault(variableName, "");
+    }
+
+    private static String replaceNonPrefix(
+            final Matcher matcher,
+            final LinkedHashMap<String, Map<String, String>> variables,
+            final ImmutableList.Builder<String> variablesUsedBuilder,
+            final String variableName) {
+        // First try the capture group with the name
+        try {
+            return matcher.group(variableName);
```

Comment:
If you are intent on keeping the numeric representation, I think that you need to push it down to this level to ensure it's used in the correct precedence order.

Reply:
Agreed. I didn't consider the case where an environment var, for example, was a number.